File: java/src/main/java/com/ytta/v0/YttaV0.java
package com.ytta.v0;

import com.ytta.v0.LatencyProbe.Summary;
import com.ytta.v0.Pipeline.GoldenEvent;
import com.ytta.v0.Pipeline.Mode;
import com.ytta.v0.Types.Tick;

import java.io.BufferedWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;

public final class YttaV0 {
    public static void main(String[] args) throws IOException {
        String fixture = null;
        String outPath = null;
        Mode mode = Mode.SYNC;
        for (int i = 0; i < args.length; i++) {
            String arg = args[i];
            if (arg.equals("--fixture")) {
                if (i + 1 >= args.length) {
                    usage();
                }
                fixture = args[++i];
            } else if (arg.equals("--out")) {
                if (i + 1 >= args.length) {
                    usage();
                }
                outPath = args[++i];
            } else if (arg.equals("--mode")) {
                if (i + 1 >= args.length) {
                    usage();
                }
                mode = parseMode(args[++i]);
            } else if (arg.startsWith("--mode=")) {
                mode = parseMode(arg.substring("--mode=".length()));
            } else if (arg.equals("--help") || arg.equals("-h")) {
                usage();
                return;
            } else {
                usage();
            }
        }
        if (fixture == null || outPath == null) {
            usage();
        }

        List<Tick> ticks = TickSource.load(Path.of(fixture));
        LatencyProbe probe = new LatencyProbe();
        List<GoldenEvent> events = Pipeline.run(ticks, probe, mode);

        try (BufferedWriter writer = Files.newBufferedWriter(Path.of(outPath))) {
            for (GoldenEvent e : events) {
                writer.write(e.line());
                writer.newLine();
            }
        }

        Summary s = probe.summarize();
        System.err.printf(
                "{\"type\":\"latency\",\"count\":%d,\"p50_ns\":%d,\"p99_ns\":%d,"
                        + "\"ingest_p50_ns\":%d,\"ingest_p99_ns\":%d,"
                        + "\"decide_p50_ns\":%d,\"decide_p99_ns\":%d,"
                        + "\"execute_p50_ns\":%d,\"execute_p99_ns\":%d,\"drops\":%d}%n",
                s.count(),
                s.p50Ns(),
                s.p99Ns(),
                s.ingestP50Ns(),
                s.ingestP99Ns(),
                s.decideP50Ns(),
                s.decideP99Ns(),
                s.executeP50Ns(),
                s.executeP99Ns(),
                s.drops());

        if (s.drops() != 0) {
            System.err.println("error: queue drops=" + s.drops());
            System.exit(1);
        }
    }

    private static Mode parseMode(String s) {
        return switch (s.toLowerCase()) {
            case "sync" -> Mode.SYNC;
            case "queued" -> Mode.QUEUED;
            default -> {
                System.err.println("error: invalid --mode " + s + " (use sync|queued)");
                System.exit(1);
                yield Mode.SYNC;
            }
        };
    }

    private static void usage() {
        System.err.println(
                "Usage: com.ytta.v0.YttaV0 --fixture <ticks.ndjson> --out <events.ndjson> [--mode=sync|queued]");
        System.exit(1);
    }
}

File: rust/src/bin/ytta_v0.rs
use std::env;
use std::fs::File;
use std::io::Write;
use std::process;

use ytta_rust::v0::{load_ticks, run_pipeline_mode, LatencyProbe, PipelineMode};

fn usage(argv0: &str) {
    eprintln!("Usage: {argv0} --fixture <ticks.ndjson> --out <events.ndjson> [--mode=sync|queued]");
}

fn parse_mode(s: &str) -> Result<PipelineMode, String> {
    match s.to_ascii_lowercase().as_str() {
        "sync" => Ok(PipelineMode::Sync),
        "queued" => Ok(PipelineMode::Queued),
        other => Err(format!("invalid --mode {other} (use sync|queued)")),
    }
}

fn main() {
    let mut fixture = None;
    let mut out_path = None;
    let mut mode = PipelineMode::Sync;
    let mut args = env::args().skip(1);
    while let Some(arg) = args.next() {
        match arg.as_str() {
            "--fixture" => fixture = args.next(),
            "--out" => out_path = args.next(),
            "--mode" => {
                let Some(v) = args.next() else {
                    usage("ytta_v0");
                    process::exit(1);
                };
                match parse_mode(&v) {
                    Ok(m) => mode = m,
                    Err(e) => {
                        eprintln!("error: {e}");
                        process::exit(1);
                    }
                }
            }
            a if a.starts_with("--mode=") => match parse_mode(&a[7..]) {
                Ok(m) => mode = m,
                Err(e) => {
                    eprintln!("error: {e}");
                    process::exit(1);
                }
            },
            "--help" | "-h" => {
                usage("ytta_v0");
                process::exit(0);
            }
            _ => {
                usage("ytta_v0");
                process::exit(1);
            }
        }
    }
    let (Some(fixture), Some(out_path)) = (fixture, out_path) else {
        usage("ytta_v0");
        process::exit(1);
    };

    let ticks = match load_ticks(&fixture) {
        Ok(t) => t,
        Err(e) => {
            eprintln!("error: {e}");
            process::exit(1);
        }
    };

    let mut probe = LatencyProbe::new();
    let events = run_pipeline_mode(&ticks, &mut probe, mode);

    let mut file = match File::create(&out_path) {
        Ok(f) => f,
        Err(e) => {
            eprintln!("error: {e}");
            process::exit(1);
        }
    };
    for e in &events {
        if writeln!(file, "{}", e.line).is_err() {
            eprintln!("error: write failed");
            process::exit(1);
        }
    }

    let s = probe.summarize();
    eprintln!(
        "{{\"type\":\"latency\",\"count\":{},\"p50_ns\":{},\"p99_ns\":{},\"ingest_p50_ns\":{},\"ingest_p99_ns\":{},\"decide_p50_ns\":{},\"decide_p99_ns\":{},\"execute_p50_ns\":{},\"execute_p99_ns\":{},\"drops\":{}}}",
        s.count,
        s.p50_ns,
        s.p99_ns,
        s.ingest_p50_ns,
        s.ingest_p99_ns,
        s.decide_p50_ns,
        s.decide_p99_ns,
        s.execute_p50_ns,
        s.execute_p99_ns,
        s.drops
    );

    if s.drops != 0 {
        eprintln!("error: queue drops={}", s.drops);
        process::exit(1);
    }
}
